streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java

```diff
@@ -632,8 +632,18 @@ public final void addInternalTopic(final String topicName,
         internalTopicNamesWithProperties.put(topicName, internalTopicProperties);
     }
 
-    public final void copartitionSources(final Collection<String> sourceNodes) {
-        copartitionSourceGroups.add(Collections.unmodifiableSet(new HashSet<>(sourceNodes)));
+    public final synchronized void copartitionSources(final Collection<String> sourceNodes) {
+        copartitionSourceGroups.add(new HashSet<>(sourceNodes));
+    }
+
+    public final synchronized void maybeUpdateCopartitionSourceGroups(final String replacedNodeName,
```

Comment:
I'm not really sure what's the threading model is when building the topology, but chose to be safe and made both accessors synchronized.